### PR TITLE
StormScope MeteoSat Europe

### PR DESCRIPTION
**Pull request by jleinonen**
# Earth2Studio Pull Request

## Description

Adds the European StormScope MeteoSat model. 

1. Model wrapper: `earth2studio.models.px.stormscope_meteosat.StormScopeMeteosatEU`
2. Example in `examples/04_nowcasting/05_stormscope_meteosat_example.py`
3. Unit tests
4. Some fixes/improvements to the `MeteosatFCI` data source:
    - Fix `projection_extent` to account for FCI coordinate convention
    - `available_variables` to list the variables available at the selected resolution
    - Recycle opened data files (this speeds up data loading in the example by ~5x on my system)

Depends on https://github.com/NVIDIA/physicsnemo/pull/1848.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.
- [x] Assess and address Greptile feedback (AI code review bot for guidance; use discretion, addressing all feedback is not required).

### CODE REVIEW

File: earth2studio/models/px/stormscope_meteosat.py [hunk 1]
```diff
@@ -0,0 +1,807 @@
...
+            lat = torch.from_numpy(metadata["lat"].values)
+            lon = torch.from_numpy(metadata["lon"].values)
+            earth_mask = torch.from_numpy(metadata["earth_mask"].values)
+            mtg_y = metadata["mtg_y"].values
+            mtg_x = metadata["mtg_x"].values
+
+        # EDM sampler arguments
+        sampler_args = config.sampler_args
+        inference_mtg_box = (
+            tuple(config.get("mtg_ylim", cls.Model_FCI_BBox[0])),
```
[the remark this answers is not in the document]
Comment (jleinonen) [reply earth2studio/models/px/stormscope_meteosat.py:865]: Lines 662-665 are supposed to fill in the angles for step L. Are you sure it's not filled?

File: earth2studio/models/px/stormscope_meteosat.py [hunk 2]
```diff
@@ -0,0 +1,807 @@
...
+    from physicsnemo.diffusion.noise_schedulers import EDMNoiseScheduler
+    from physicsnemo.diffusion.preconditioners import EDMPreconditioner
+    from physicsnemo.diffusion.samplers import sample
+    from physicsnemo.utils.zenith_angle import zenith_azimuth_angles
+except ImportError:
+    OptionalDependencyFailure("stormscope-meteosat")
+    OmegaConf = None
+
+
+# Variables used in StormCastMTG
```
[the remark this answers is not in the document]
Comment (jleinonen) [reply earth2studio/models/px/stormscope_meteosat.py]: Fixed in https://github.com/NVIDIA/earth2studio/pull/981/commits/447141b9f9c5c738553fd965296ec919c5df5a33

File: earth2studio/models/px/stormscope_meteosat.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jleinonen) [reply earth2studio/models/px/stormscope_meteosat.py:865]: Fixed in https://github.com/NVIDIA/earth2studio/pull/981/commits/6e36c0c40b3f8232732944278a7b33d626686ab2

File: earth2studio/models/px/stormscope_meteosat.py [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+        Output time offsets relative to the analysis time; single-step prediction
+        only, by default ``[10 min]``
+    ir_38_warm_scale_factor : float, optional
+        Multiplier applied to the above-threshold branch of the ir_38 channel after
+        raw-to-physical conversion, by default 0.024222141
+    ir_38_warm_threshold : float, optional
+        Raw digital-count boundary (the 12-bit maximum) above which the warm/HDR
+        ir_38 scaling is applied, by default 4095.0
+    num_diffusion_steps : int, optional
+        Number of EDM diffusion sampling steps, by default 18
```
[the remark this answers is not in the document]
Comment (jleinonen) [reply earth2studio/models/px/stormscope_meteosat.py]: Fixed in https://github.com/NVIDIA/earth2studio/pull/981/commits/5095bb8fb409c9ca4c6dd7aeaeeeff1db5d5aa1e

File: examples/04_nowcasting/04_stormscope_meteosat_example.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    x = torch.nn.functional.avg_pool2d(x.reshape(prod(batch_dims), *x.shape[-3:]), 2)
+    x = x.reshape(*batch_dims, *x.shape[-3:])
+
+    # merge downsampled and native 2km data
+    x = torch.concat([x, x_res["2km"][0]], dim=-3)
+    coords = x_res["2km"][1]
+    coords["variable"] = np.concatenate(
+        [x_res["1km"][1]["variable"], coords["variable"]]
+    )
+    del x_res
```
Comment (pzharrington) [inline examples/04_nowcasting/04_stormscope_meteosat_example.py:172]: What is going on here with all the 2km vs. 1km data stuff? It's not explained in the comment block above this cell

File: examples/04_nowcasting/04_stormscope_meteosat_example.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+# Load Model
+# ----------
+# Load StormScopeMeteosatEU from a package.
+
+# %%
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+# Load the model package from a local path or remote URI.
+# Set STORMSCOPE_METEOSAT_MODEL_PATH to override the default location.
+model_path = os.environ.get("STORMSCOPE_METEOSAT_MODEL_PATH")
```
Comment (pzharrington) [inline examples/04_nowcasting/04_stormscope_meteosat_example.py]: Remove this and just rely on `load_default_package` which should point to HF

File: earth2studio/models/px/stormscope_meteosat.py [hunk 6]
```diff
@@ -0,0 +1,807 @@
...
+        x = self.raw_to_normalized(x, in_place=True)
+        x.masked_fill_(self.off_earth_mask_tensor, 0)
+        return x
+
+    def denormalize(self, x: torch.Tensor, in_place: bool = False) -> torch.Tensor:
+        """Denormalise ``x`` back to physical units."""
+        x = self.normalized_to_raw(x, in_place=in_place)
+        x = self.raw_to_physical(x, in_place=True)
+        x.masked_fill_(self.off_earth_mask_tensor, torch.nan)
+        return x
```
Comment (pzharrington) [inline earth2studio/models/px/stormscope_meteosat.py:440]: Three different data representations here -- physical -> raw -> normalized. Typically we just have physical and normalized, wondering what is up with the extra tier...can you explain these in some more detail? Mainly trying to decide whether it might make sense to migrate some of these data unit handlers as internal helpers or public classmethods of the FCI data source if they might be used more generally

File: earth2studio/models/px/stormscope_meteosat.py [hunk 7]
```diff
@@ -0,0 +1,807 @@
...
+        condition = torch.cat(
+            (zen_azi, x, invariant_tensor),
+            dim=1,
+        )
+        denoiser = self.scheduler.get_denoiser(
+            x0_predictor=self._conditioned_x0_predictor(condition)
+        )
+        return self._sample(
+            denoiser, shape=(B, C, H, W), dtype=x.dtype, device=x.device
+        )
```
Comment (pzharrington) [inline earth2studio/models/px/stormscope_meteosat.py:536]: ~Single expert denoising?~ nvm, see the multi-expert clause in the x0_predictor closure

File: earth2studio/models/px/stormscope_meteosat.py [hunk 8]
```diff
@@ -0,0 +1,807 @@
...
+        """Load prognostic package"""
+        # package = Package(
+        #     "hf://nvidia/stormcast-conus@...",
+        #     cache_options={
+        #         "cache_storage": Package.default_cache("stormcast-conus"),
+        #         "same_names": True,
+        #     },
+        # )
+        # return package
+        raise NotImplementedError("No default package.")
```
Comment (pzharrington) [inline earth2studio/models/px/stormscope_meteosat.py]: Let's point this to HF once I upload

File: earth2studio/models/px/stormscope_meteosat.py [hunk 9]
```diff
@@ -0,0 +1,807 @@
...
+
+    The model operates on a rectangular sub-region of the MTG full-disk image specified
+    by ``mtg_ylim`` and ``mtg_xlim`` in native MTG pixel coordinates.
+
+    Parameters
+    ----------
+    model_low : torch.nn.Module
+        Denoising sub-model applied at low noise levels (sigma < ``sigma_threshold``).
+    model_high : torch.nn.Module
+        Denoising sub-model applied at high noise levels (sigma ≥ ``sigma_threshold``).
```
Comment (pzharrington) [inline earth2studio/models/px/stormscope_meteosat.py]: Using a `model_spec` dict like I did for StormScopeGOES/MRMS might be a little more future-proof if we have future variants with more or less than 2 experts https://github.com/NVIDIA/earth2studio/blob/f6996708ee6f24f109d665c094daad61ca71b069/earth2studio/models/px/stormscope.py#L74

File: examples/04_nowcasting/04_stormscope_meteosat_example.py (hunk 4 above)
Comment (jleinonen) [reply examples/04_nowcasting/04_stormscope_meteosat_example.py:172]: MTG visible and NIR channels are provided by the EUMETSAT data store at 1 km resolution while IR channels are 2 km. StormScope MTG uses the 2 km grid for all channels. The 1 km grid is exactly 2 times the pixels of the 2 km grid, so we can use 2x2 average pooling to produce the 2 km inputs from the 1 km data.

I'll add a mention of this in the comment block - and also maybe this could be a utility function somewhere?

File: examples/04_nowcasting/04_stormscope_meteosat_example.py (hunk 5 above)
Comment (jleinonen) [reply examples/04_nowcasting/04_stormscope_meteosat_example.py]: Yep this is just a placeholder so we can test while the model is not on HF. Will remove once model is available on HF.

File: earth2studio/models/px/stormscope_meteosat.py (hunk 6 above)
Comment (jleinonen) [reply earth2studio/models/px/stormscope_meteosat.py:440]: There are three forms of the data considered:

- **Raw**: uint16 values used in the MTG-FCI data files
- **Normalized**: The internal representation of the data used by the model. This is simply raw values shifted and scaled to (~ 0-mean, 1-std).
- **Physical**: The actual physical radiances. These are also a simple shift+scale from the raw data, except for the IR3.8 channel, which requires special handling.

The shift-and-scale coefficients can be fused except for the IR3.8 channel. So we could reduce the transformation for the other channels to a single op, but I'm not sure it's worth the complexity.

File: earth2studio/models/px/stormscope_meteosat.py (hunk 9 above)
Comment (jleinonen) [reply earth2studio/models/px/stormscope_meteosat.py]: Okay, this is now updated to use `model_spec`. The external interface is similar to `StormScopeGOES` but the internal implemention is a bit different to fit the existing structure.

File: earth2studio/models/px/stormscope_meteosat.py (hunk 6 above)
Comment (pzharrington) [reply earth2studio/models/px/stormscope_meteosat.py:440]: I guess it's just a bit odd to me "raw" tier seems to be the version the model was actually trained on. But ok, sounds like basically the data source will always return physical values (i.e. the IR3.8 transformation has already been applied) so not needed to add extra things to the data source. How about we just add a bit more explanatory comments in the wrapper here so the behavior/reasons are documented?

Comment (pzharrington) [reply earth2studio/models/px/stormscope_meteosat.py:440]: Basically just somewhere in the wrapper, define these three tiers and why the transformations need to be applied in that order

File: examples/04_nowcasting/04_stormscope_meteosat_example.py (hunk 4 above)
Comment (pzharrington) [reply examples/04_nowcasting/04_stormscope_meteosat_example.py:172]: Yeah I'm in favor of both more explanation in the comment block and a utility function. Probably makes more sense to have the utility function be a model classmethod since it's model-specific?

Comment (jleinonen) [reply examples/04_nowcasting/04_stormscope_meteosat_example.py:172]: Yes, that makes sense - I updated the comments and will add the classmethod soon.

File: earth2studio/models/px/stormscope_meteosat.py (hunk 6 above)
Comment (jleinonen) [reply earth2studio/models/px/stormscope_meteosat.py:440]: Improved the docs of these in https://github.com/NVIDIA/earth2studio/pull/981/commits/06bc54e14b0ccc6937c4fd9e4c401395a2424f39.

File: examples/04_nowcasting/04_stormscope_meteosat_example.py (hunk 4 above)
Comment (jleinonen) [reply examples/04_nowcasting/04_stormscope_meteosat_example.py:172]: Classmethod is added: https://github.com/NVIDIA/earth2studio/pull/981/commits/c5f40004bb66a7f86d477fab85497b3df6d7d586.

**Comment by pzharrington** [comment]
Before merge we need to update it with upstream main branch, mainly which just landed the PR for the docs overhaul: https://github.com/NVIDIA/earth2studio/pull/1054

When you update this branch, make sure to resolve the conflicts and importantly make sure the example/docs entries here use the updated syntax for correct rendering (see the linked PR for examples). Sorry this happened while your PR was still open

File: earth2studio/models/px/stormscope_meteosat.py (hunk 8 above)
Comment (pzharrington) [reply earth2studio/models/px/stormscope_meteosat.py]: @jleinonen should be uploaded now, you can use the tag `17148d712f7ef2c0b48355032dcad85173dd230e` with https://huggingface.co/nvidia/stormscope-meteosat

File: examples/04_nowcasting/04_stormscope_meteosat_example.py (hunk 5 above)
Comment (pzharrington) [reply examples/04_nowcasting/04_stormscope_meteosat_example.py]: HF is ready now

File: examples/04_nowcasting/04_stormscope_meteosat_example.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+# Set Up Data Sources
+# -------------------
+# ``MeteosatFCI`` is configured with ``pixel_bbox`` matching the model's
+# domain (``Model_FCI_BBox``) so that the returned ``y``/``x`` scan-angle
+# coordinates align #with what the model expects.
+#
+# Meteosat FCI visible and NIR channels are provided by the EUMETSAT data store
+# at 1 km resolution while IR channels are 2 km, so we set up data sources for
+# both. The 1 km grid is exactly 2 times the pixels of the 2 km grid, so we get
+# the 1 km bounding box by doubling the 2 km bounding box.
```
Comment (pzharrington) [inline examples/04_nowcasting/04_stormscope_meteosat_example.py:118]: I suggest the following to make it more clear:
```
# Meteosat FCI visible and NIR channels are provided by the EUMETSAT data store
# at 1 km resolution while IR channels are 2 km. The StormScope MTG model uses
# the 2 km grid for all channels, so we need to pull data from both 1 km and 2 km 
# sources and do a minimal regrid on the visible and NIR channels.

# Here we will set up data sources for both grid resolutions.
# The 1 km grid is exactly 2 times the pixels of the 2 km grid, so we get
# the 1 km bounding box by doubling the 2 km bounding box.
```

**Review by pzharrington — APPROVED** [review]
